core/dbt/parser/partial.py

```diff
@@ -32,10 +33,21 @@
 }
 
 
+# These macro names have special treatment in the ManifestLoader and
+# partial parsing. If they have changed we will skip partial parsing
+special_override_macros = [
+    'ref', 'source', 'config', 'generate_schema_name',
+    'generate_database_name', 'generate_alias_name'
+]
+
+
 # Partial parsing. Create a diff of files from saved manifest and current
 # files and produce a project_parser_file dictionary to drive parsing of
 # only the necessary changes.
 # Will produce a 'skip_parsing' method, and a project_parser_file dictionary
+# All file objects from the new manifest are deepcopied, because we need
+# to preserve an unchanged file object in case we need to drop back to a
+# a full parse (such as for certain macro changes)
```

Comment:
We're copying the sourcefile. We need to leave all of the information in the file objects unchanged because they need to be unchanged if we need to restore the the originally read files because of aborting the partial parse. The other alternative would be to use to_dict() and restore from a dictionary, but to_dict() on files doesn't have the contents.

Reply:
What's the worst case scenario for this then? Do we end up deep copying all source files in a project?

Reply:
Talked with @gshank and it looks like these calls to `deepcopy` only happen on the changed files so there won't be very many calls to them. I'd like to be able to bypass these copies and trigger a reparse directly after parsing macros, but there's enough mutation to the nodes to schedule parsing that we can't easily shortcut at that stage.

To me, it looks like the performance impact here is acceptable since it is proportional to the number of files changed which should be expected for users.